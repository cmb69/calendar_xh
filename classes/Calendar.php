<?php

/**
 * Copyright 2021 Christoph M. Becker
 *
 * This file is part of Calendar_XH.
 *
 * Calendar_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Calendar_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Calendar_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Calendar;

class Calendar
{
    /** @var bool */
    private $weekStartsMonday;

    /**
     * @param bool $weekStartsMonday
     */
    public function __construct($weekStartsMonday)
    {
        $this->weekStartsMonday = $weekStartsMonday;
    }

    /**
     * @param int $year
     * @param int $month
     * @return array<int,array<int,(int|null)>>
     */
    public function getMonthMatrix($year, $month)
    {
        $result = [];
        $row = [];
        $days = (int) date('t', mktime(1, 1, 1, $month, 1, $year));
        for ($day = 1; $day <= $days; $day++) {
            $dayofweek = $this->getDayOfWeek($year, $month, $day);
            if ($day === 1) {
                for ($i = 0; $i < $dayofweek; $i++) {
                    $row[] = null;
                }
            }
            $row[] = $day;
            if ($day === $days) {
                for ($i = $dayofweek + 1; $i < 7; $i++) {
                    $row[] = null;
                }
            }
            if (count($row) === 7 || $day === $days) {
                $result[] = $row;
                $row = [];
            }
        }
        return $result;
    }

    /**
     * @param int $year
     * @param int $month
     * @param int $day
     * @return int
     */
    private function getDayOfWeek($year, $month, $day)
    {
        $dayofweek = (int) date('w', mktime(1, 1, 1, $month, $day, $year));
        if ($this->weekStartsMonday) {
            $dayofweek = ($dayofweek + 6) % 7;
        }
        return $dayofweek;
    }
}
