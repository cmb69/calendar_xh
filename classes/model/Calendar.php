<?php

/**
 * Copyright (c) Christoph M. Becker
 *
 * This file is part of Calendar_XH.
 *
 * Calendar_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Calendar_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Calendar_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Calendar\Model;

use Calendar\Html2Text;

class Calendar
{
    /** @var array<string,Event> */
    private $events;

    /** @param array<string,Event> $events */
    public function __construct(array $events)
    {
        $this->events = $events;
    }

    /** @return array<string,Event> */
    public function events(): array
    {
        return $this->events;
    }

    public function event(string $id): ?Event
    {
        return $this->events[$id] ?? null;
    }

    public function nextEvent(LocalDateTime $now): ?Event
    {
        $nextevent = null;
        $nextldt = null;
        foreach ($this->events as $event) {
            $ldt = $event->after($now);
            if ($ldt === null) {
                continue;
            }
            if ($nextldt === null || $ldt->compare($nextldt) < 0) {
                $nextevent = $event;
                $nextldt = $ldt;
            }
        }
        return $nextevent;
    }

    /** @return list<Event> */
    public function eventsDuring(int $year, int $month): array
    {
        $result = [];
        foreach ($this->events as $event) {
            if ($event->occursDuring($year, $month)) {
                $result[] = $event;
            }
        }
        uasort($result, function (Event $a, Event $b) use ($year): int {
            $dt1 = $a->isBirthday() ? $a->start()->withYear($year) : $a->start();
            $dt2 = $b->isBirthday() ? $b->start()->withYear($year) : $b->start();
            return $dt1->compare($dt2);
        });
        return $result;
    }

    /** @return list<Event> */
    public function eventsOn(LocalDateTime $day, bool $daysBetween): array
    {
        assert($day->hour() === 0 && $day->minute() === 0);
        $result = [];
        foreach ($this->events as $event) {
            if ($event->occursOn($day, $daysBetween)) {
                $result[] = $event;
            }
        }
        return $result;
    }

    public function delete(string $id): void
    {
        assert(array_key_exists($id, $this->events));
        unset($this->events[$id]);
    }

    public function toICalendarString(Html2Text $converter, string $host): string
    {
        $res = "BEGIN:VCALENDAR\r\n"
            . "PRODID:-//3-magi.net//Calendar_XH//EN\r\n"
            . "VERSION:2.0\r\n";
        foreach ($this->events() as $id => $event) {
            $res .= $event->toICalendarString($id, $converter, $host);
        }
        $res .= "END:VCALENDAR\r\n";
        return $res;
    }
}
