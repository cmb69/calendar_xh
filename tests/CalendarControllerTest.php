<?php

/**
 * Copyright 2023 Christoph M. Becker
 *
 * This file is part of Calendar_XH.
 *
 * Calendar_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Calendar_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Calendar_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Calendar;

use ApprovalTests\Approvals;
use Calendar\Model\Calendar;
use Calendar\Model\Event;
use PHPUnit\Framework\TestCase;
use Plib\FakeRequest;
use Plib\View;

class CalendarControllerTest extends TestCase
{
    public function testDefaultActionRendersHtml()
    {
        $plugin_cf = XH_includeVar("./config/config.php", 'plugin_cf');
        $conf = $plugin_cf['calendar'];
        $lang = XH_includeVar("./languages/en.php", "plugin_tx")["calendar"];
        $eventDataService = $this->createStub(EventDataService::class);
        $eventDataService->method("readEvents")->willReturn(
            new Calendar([$this->lunchBreak(), $this->weekend(), $this->birthday()])
        );
        $dateTimeFormatter = new DateTimeFormatter($lang);

        $view = new View("./views/", $lang);
        $sut = new CalendarController(
            "./",
            $conf,
            $eventDataService,
            $dateTimeFormatter,
            1,
            new Counter(1),
            $view
        );
        $request = new FakeRequest([
            "url" => "http://example.com/?page",
            "time" => 1675088820,
        ]);
        $response = $sut->defaultAction(0, 0, "", $request);
        Approvals::verifyHtml($response->output());
    }

    private function lunchBreak(): Event
    {
        return Event::create(
            "2023-01-04",
            "2023-01-04",
            "12:00",
            "13:00",
            "Lunch break",
            "http://example.com/lunchbreak",
            "Tips for lunch breaks",
            "whereever I am",
            ""
        );
    }

    private function weekend(): Event
    {
        return Event::create(
            "2023-01-07",
            "2023-01-08",
            "",
            "",
            "Weekend",
            "",
            "",
            "",
            ""
        );
    }

    private function birthday(): Event
    {
        return Event::create(
            "2000-01-01",
            "",
            "",
            "",
            "Millenium",
            "",
            "",
            "###",
            ""
        );
    }
}
