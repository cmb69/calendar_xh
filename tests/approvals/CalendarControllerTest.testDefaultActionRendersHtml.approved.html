<div class="calendar_calendar">
<script type="module" src="/js/calendar.js?&v=2.7-dev"></script>
<table class="calendar_main" role="presentation">
  <caption class="calendar_monthyear">
    <a href="/?page&amp;month=12&amp;year=2022" aria-labelledby="calendar_id_5">
      <span>◄</span>
      <span role="tooltip" id="calendar_id_5">Previous month</span>
    </a>
    <span>January 2023</span>
    <a href="/?page&amp;month=2&amp;year=2023" aria-labelledby="calendar_id_6">
      <span>►</span>
      <span role="tooltip" id="calendar_id_6">Next month</span>
    </a>
  </caption>
  <tr>
    <th class="calendar_daynames calendar_day" scope="col">
      <abbr title="Monday">Mo</abbr>
    </th>
    <th class="calendar_daynames calendar_day" scope="col">
      <abbr title="Tuesday">Tu</abbr>
    </th>
    <th class="calendar_daynames calendar_day" scope="col">
      <abbr title="Wednesday">We</abbr>
    </th>
    <th class="calendar_daynames calendar_day" scope="col">
      <abbr title="Thursday">Th</abbr>
    </th>
    <th class="calendar_daynames calendar_day" scope="col">
      <abbr title="Friday">Fr</abbr>
    </th>
    <th class="calendar_daynames calendar_we" scope="col">
      <abbr title="Saturday">Sa</abbr>
    </th>
    <th class="calendar_daynames calendar_we" scope="col">
      <abbr title="Sunday">Su</abbr>
    </th>
  <tr>
  <tr>
    <td class="calendar_noday">
        <span></span>
  </td>
    <td class="calendar_noday">
        <span></span>
  </td>
    <td class="calendar_noday">
        <span></span>
  </td>
    <td class="calendar_noday">
        <span></span>
  </td>
    <td class="calendar_noday">
        <span></span>
  </td>
    <td class="calendar_noday">
        <span></span>
  </td>
    <td class="calendar_eventday calendar_we">
      <a href="/?Events&amp;month=1&amp;year=2023" aria-describedby="calendar_id_1")>
        <span>1</span>
        <span role="tooltip" id="calendar_id_1">Millenium 23 years</span>
      </a>
    </td>
  </tr>
  <tr>
    <td class="calendar_day">
        <span>2</span>
  </td>
    <td class="calendar_day">
        <span>3</span>
  </td>
    <td class="calendar_eventday calendar_day">
      <a href="/?Events&amp;month=1&amp;year=2023" aria-describedby="calendar_id_2")>
        <span>4</span>
        <span role="tooltip" id="calendar_id_2">12:00 pm - 1:00 pm Lunch break</span>
      </a>
    </td>
    <td class="calendar_day">
        <span>5</span>
  </td>
    <td class="calendar_day">
        <span>6</span>
  </td>
    <td class="calendar_eventday calendar_eventstart calendar_we">
      <a href="/?Events&amp;month=1&amp;year=2023" aria-describedby="calendar_id_3")>
        <span>7</span>
        <span role="tooltip" id="calendar_id_3">12:00 am Weekend till 1/8/2023 11:59 pm</span>
      </a>
    </td>
    <td class="calendar_eventday calendar_eventend calendar_we">
      <a href="/?Events&amp;month=1&amp;year=2023" aria-describedby="calendar_id_4")>
        <span>8</span>
        <span role="tooltip" id="calendar_id_4">12:00 am Weekend till 1/8/2023 11:59 pm</span>
      </a>
    </td>
  </tr>
  <tr>
    <td class="calendar_day">
        <span>9</span>
  </td>
    <td class="calendar_day">
        <span>10</span>
  </td>
    <td class="calendar_day">
        <span>11</span>
  </td>
    <td class="calendar_day">
        <span>12</span>
  </td>
    <td class="calendar_day">
        <span>13</span>
  </td>
    <td class="calendar_we">
        <span>14</span>
  </td>
    <td class="calendar_we">
        <span>15</span>
  </td>
  </tr>
  <tr>
    <td class="calendar_day">
        <span>16</span>
  </td>
    <td class="calendar_day">
        <span>17</span>
  </td>
    <td class="calendar_day">
        <span>18</span>
  </td>
    <td class="calendar_day">
        <span>19</span>
  </td>
    <td class="calendar_day">
        <span>20</span>
  </td>
    <td class="calendar_we">
        <span>21</span>
  </td>
    <td class="calendar_we">
        <span>22</span>
  </td>
  </tr>
  <tr>
    <td class="calendar_day">
        <span>23</span>
  </td>
    <td class="calendar_day">
        <span>24</span>
  </td>
    <td class="calendar_day">
        <span>25</span>
  </td>
    <td class="calendar_day">
        <span>26</span>
  </td>
    <td class="calendar_day">
        <span>27</span>
  </td>
    <td class="calendar_we">
        <span>28</span>
  </td>
    <td class="calendar_we">
        <span>29</span>
  </td>
  </tr>
  <tr>
    <td class="calendar_today calendar_day">
        <span>30</span>
  </td>
    <td class="calendar_day">
        <span>31</span>
  </td>
    <td class="calendar_noday">
        <span></span>
  </td>
    <td class="calendar_noday">
        <span></span>
  </td>
    <td class="calendar_noday">
        <span></span>
  </td>
    <td class="calendar_noday">
        <span></span>
  </td>
    <td class="calendar_noday">
        <span></span>
  </td>
  </tr>
</table>
</div>